third_party/intel/lib/TritonGENToLLVM/Attributes.cpp

```diff
@@ -0,0 +1,436 @@
+  bool ContainsPassThroughAttr =
+      getAttribute(AttributeList::PassThroughAttrName).has_value();
+  if (ContainsPassThroughAttr) {
+    NamedAttribute PassThroughAttr =
+        getAttribute(AttributeList::PassThroughAttrName).value();
+    auto ArrAttr = cast<ArrayAttr>(PassThroughAttr.getValue());
+    std::vector<mlir::Attribute> Vec = ArrAttr.getValue().vec();
+
+    std::remove_if(Vec.begin(), Vec.end(), [AttrName](mlir::Attribute &Attr) {
+      if (isa<StringAttr>(Attr))
+        return (cast<StringAttr>(Attr).strref() == AttrName);
+      if (isa<ArrayAttr>(Attr)) {
+        auto ArrAttr = cast<ArrayAttr>(Attr);
+        assert(ArrAttr.size() == 2 && isa<StringAttr>(ArrAttr[0]));
+        return (cast<StringAttr>(ArrAttr[0]).strref() == AttrName);
+      }
+      return false;
+    });
+
+    PassThroughAttr.setValue(ArrayAttr::get(&Ctx, Vec));
+    return *this;
+  }
+
+  Attrs.erase(AttrName);
+  return *this;
+}
+
+AttrBuilder &AttrBuilder::removeAttribute(llvm::Attribute::AttrKind Kind) {
+  assert((unsigned)Kind < llvm::Attribute::EndAttrKinds &&
+         "Attribute out of range!");
+  StringRef AttrName = llvm::Attribute::getNameFromAttrKind(Kind);
+  return removeAttribute(AttrName);
+}
+
+bool AttrBuilder::contains(StringRef AttrName) const {
+  if (containsInPassThrough(AttrName))
+    return true;
+  return getAttribute(AttrName).has_value();
+}
+
+bool AttrBuilder::contains(llvm::Attribute::AttrKind Kind) const {
+  StringRef AttrName = llvm::Attribute::getNameFromAttrKind(Kind);
+  return contains(AttrName);
+}
+
+std::optional<NamedAttribute>
+AttrBuilder::getAttribute(StringRef AttrName) const {
+  return Attrs.getNamed(AttrName);
+}
+
+std::optional<NamedAttribute>
+AttrBuilder::getAttribute(llvm::Attribute::AttrKind Kind) const {
+  StringRef AttrName = llvm::Attribute::getNameFromAttrKind(Kind);
+  return getAttribute(AttrName);
+}
+
+StringAttr AttrBuilder::createStringAttribute(
+    Twine AttrName, std::optional<StringLiteral> Prefix, MLIRContext &Ctx) {
+  return (Prefix) ? StringAttr::get(&Ctx, *Prefix + "." + AttrName)
+                  : StringAttr::get(&Ctx, AttrName);
+}
+
+AttrBuilder &AttrBuilder::addAttributeImpl(llvm::Attribute::AttrKind Kind,
+                                           std::optional<StringLiteral> Dialect,
+                                           AddAttrFuncPtr AddAttrPtr) {
+  assert(AddAttrPtr && "'AddAttrPtr' should be a valid function pointer");
+
+  // TODO: Replace with std::invoke once C++17 headers are available.
+  auto Invoke = [this](AddAttrFuncPtr AddAttrPtr,
+                       auto... Args) -> AttrBuilder & {
+    return (this->*AddAttrPtr)(Args...);
+  };
+
+  OpBuilder Builder(&Ctx);
+  StringRef AttrName = llvm::Attribute::getNameFromAttrKind(Kind);
+  NamedAttribute NamedAttr(createStringAttribute(AttrName, Dialect, Ctx),
+                           Builder.getUnitAttr());
+  return Invoke(AddAttrPtr, NamedAttr);
```

Comment:
C++ 17 headers are available already. However, if we use the `llvm::function_ref`, we should be able to just :
```suggestion
  OpBuilder Builder(&Ctx);
  StringRef AttrName = llvm::Attribute::getNameFromAttrKind(Kind);
  NamedAttribute NamedAttr(createStringAttribute(AttrName, Dialect, Ctx),
                           Builder.getUnitAttr());
  return AddAttrPtr(*this, NamedAttr);
```

Reply:
Applies to all occurrences below.

Reply:
using C++17 std::invoke

Reply:
using std::invoke for all